lib/bq_stream/archive.rb

```diff
@@ -0,0 +1,136 @@
+            # limited to the number of records we can grab, keeping under 10_000 rows
+            # MUST use find_by_sql to avoid any default scopes (looking at you JobLogRecord)
+            next_batch = table_class.unscoped.order(id: :desc).where('id >= ? AND id <= ?', back_date_id, earliest_record_id).limit(10_000 / (oldest_attr_recs.count.zero? ? 1 : oldest_attr_recs.count)) rescue []
+            BqStream.log(:info, "#{Time.now}: ***** Next Batch Count for #{table}: #{next_batch.count} *****")
+
+            if next_batch.empty?
+              # If there are no records in range mark the table's attributes as archived
+              BqStream::OldestRecord.where(table_name: table).each { |row| row.update(archived: true) }
+            else
+              # Write rows from records for BigQuery and place them into the buffer 
+              oldest_attr_recs.uniq.each do |oldest_attr_rec|
+                next_batch.each do |record|
+                  new_val = record[oldest_attr_rec.attr] && table.constantize.type_for_attribute(oldest_attr_rec.attr).type == :datetime ? record[oldest_attr_rec.attr].in_time_zone(time_zone) : record[oldest_attr_rec.attr].to_s
+                  buffer << { table_name: table,
+                              record_id: record.id,
+                              attr: oldest_attr_rec.attr,
+                              new_value: new_val,
+                              updated_at: record.created_at }
+                end
+              end
+            end
+
+            # Only write to BigQuery if we have rows set in the buffer
+            unless buffer.empty?
+              BqStream.log(:info, "#{Time.now}: ***** Start data pack and insert *****")
+              # Create data object for BigQuery insert
+              data = buffer.collect do |i|
+                new_val = encode_value(i[:new_value]) rescue nil
+                { table_name: i[:table_name], record_id: i[:record_id], attr: i[:attr],
+                  new_value: new_val ? new_val : i[:new_value], updated_at: i[:updated_at] }
+              end
+              BqStream.log(:info, "#{Time.now}: ***** Insert #{buffer.count} Records for #{table} to BigQuery *****")
+              # Insert rows into BigQuery in one call
+              @bq_archiver.insert(bq_table_name, data) unless data.empty?
+              # Lower earliest written id by one to get the next record to be written
+              earliest_record_id = next_batch.map(&:id).min - 1
+              BqStream.log(:info, "#{Time.now}: ***** updated earliest_record_id: #{earliest_record_id} *****")
+              # Clear the buffer for next cycle through the table
+              buffer = []
+              BqStream.log(:info, "#{Time.now}: ***** End data pack and insert *****")
+            end
+          end
+        end
+      end
+      BqStream.log(:info, "#{Time.now}: ***** End Streamline Process *****")
+    end
+  end
+end
```

Comment:
There seems to be some duplication between this new module and the existing code, and this is literally one giant method in a file. I want to just get this out there and done, so put in an issue to refactor and clean this up but let's get this merged now and over with.